tf_agents/networks/multi_categorical_projection_network.py

```diff
@@ -0,0 +1,140 @@
+"""Project inputs to multiple categorical distribution object."""
+
+import gin
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from tf_agents.networks import network
+from tf_agents.networks import utils
+from tf_agents.specs import tensor_spec
+from tf_agents.specs import distribution_spec
+
+class MultiCategoricalDistributionBlock(tfp.distributions.Blockwise):
+
+  def __init__(self, logits, categories_shape):
+    self.categories_shape = categories_shape
+    distribs = self._create_distrib(logits)
+    super().__init__(distributions = distribs)
+    self._parameters['logits'] = logits
+
+  def _create_distrib(self, logits):
+    logits_splitted = tf.split(logits, self.categories_shape, -1)
+    distribs = tf.nest.map_structure(
+                    lambda l: tfp.distributions.Categorical(logits = l),
+                    logits_splitted)
+    return distribs
+
+  def _mode(self):
+    return self._flatten_and_concat_event(
+      self._distribution.mode()
+      )
+
+@gin.configurable
+class MultiCategoricalProjectionNetwork(network.DistributionNetwork):
+  """Generates a set of tfp.distribution.Categorical by predicting logits"""
+  def __init__(self,
+            sample_spec,
+            logits_init_output_factor=0.1,
+            name='MultiCategoricalProjectionNetwork'):
+    """Creates an instance of MultiCategoricalProjectionNetwork
+
+    Args:
+      sample_spec: A collection of `tensor_spec.BoundedTensorSpec` detailing
+        the shape and dtypes of samples pulled from the output distribution.
+      logits_init_output_factor: Output factor for initializing kernal
+        logits weights.
+      name: A string representing the name of the network.
+    """
+
+    categories_shape = self._categories_shape(sample_spec)
+    n_unique_categories = np.sum(categories_shape)
+    output_spec = self._output_distribution_spec(
+      [n_unique_categories],
+      sample_spec,
+      categories_shape,
+      name)
+
+    super(MultiCategoricalProjectionNetwork, self).__init__(
+      input_tensor_spec=None,
+      state_spec = (),
+      output_spec=output_spec,
+      name=name
+    )
+
+    self._sample_spec = sample_spec
+    self.categories_shape = categories_shape
+    self.n_unique_categories = n_unique_categories
+
+    self._projection_layer = tf.keras.layers.Dense(
+        self.n_unique_categories,
+        kernel_initializer=tf.compat.v1.keras.initializers.VarianceScaling(
+            scale=logits_init_output_factor),
+        bias_initializer=tf.keras.initializers.Zeros(),
+        name='logits'
+        )
+
+  def _categories_shape(self, sample_spec):
+    def _get_n_categories(array_spec):
+      if not tensor_spec.is_bounded(array_spec):
```

Comment:
Simplify to
if tensor_spec.is_bounded(array_spec) and tensor_spec.is_discrete(array_spec) :
  n_categories = array_spec.maximum - array_spec.minimum + 1
  return n_categories
else:
  raise ValueError('sample_spec must be discrete and bounded. Got: %s.' % array_spec)